### DISCUSSION TITLE
Adding Custom CSP into exported NextJs sites

### DISCUSSION THREAD

**Discussion by sm1990** [community/community#64814] [Other Feature Feedback, Questions, & Ideas] [answered]
GitHub Pages

Question

### Body

Is there a way to add Custom CSP into exported NextJs sites deployed via GitHub pages?

**Comment by arafajs** [comment]
Unfortunately, there is no way to add a custom CSP into exported Next.js sites deployed via GitHub Pages. GitHub Pages does not support setting HTTP headers, which is how CSP is implemented.

**Comment by nftushar** [comment, marked as answer]
As of my last knowledge update in September 2021, GitHub Pages has certain limitations when it comes to adding custom security headers like Content Security Policy (CSP) directly to the deployed site. GitHub Pages provides limited configuration options, and adding custom security headers is not directly supported through GitHub Pages' settings.

However, if you want to add a custom CSP to your Next.js site deployed via GitHub Pages, you'll need to explore alternative methods:

1. **Deploy on a Different Platform**: Consider using other hosting platforms that provide more control over server configurations and allow you to set custom security headers. Platforms like Vercel, Netlify, or even a traditional web server would give you more flexibility in configuring security headers.

2. **Server-Side Rendering (SSR)**: If you're open to server-side rendering, you can deploy your Next.js app on a server that you control. This would allow you to configure security headers including CSP as part of your server setup.

3. **Content Delivery Network (CDN)**: You might use a CDN that allows you to add custom headers. Deploy your Next.js app on a platform that provides more advanced CDN features and allows custom header configurations.

4. **Middleware/Proxy**: You could deploy your Next.js app with a reverse proxy server (like Nginx or Express.js) in front of it. This server can be configured to add custom headers, including CSP, before passing the requests to your Next.js app.

5. **Inline CSP in HTML**: While not the best practice, you can consider adding your CSP rules inline within the `<meta>` tag in your HTML. This should be done with caution, as inline CSP can be tricky to manage and maintain.

6. **Frame your GitHub Pages**: If the primary reason for using GitHub Pages is for hosting and version control, you could deploy your app elsewhere and then embed it within a GitHub Pages site using an iframe. This way, you have control over the CSP of the host site while still utilizing GitHub Pages for the presentation layer.

**Comment by nirav-gajera** [comment]
to add a Custom CSP into exported Next.js sites deployed via GitHub Pages. To do this, you can add the following code to your `next.config.js` file:

```
const nextConfig = {
  ...
  headers: {
    // This is the custom CSP policy.
    "Content-Security-Policy": "default-src 'self'; script-src 'self' 'unsafe-inline'"
  }
}

module.exports = nextConfig;
```

This will add the following CSP policy to your site:

```
default-src 'self';
script-src 'self' 'unsafe-inline'
```
